File: haskell/glue/src/Glue/Lib/Bool/Until.hs
module Glue.Lib.Bool.Until where

import Glue.Eval (Eval, eval, throwError)
import Glue.Eval.Exception (wrongArgumentType)
import Glue.IR (IR (..))

until_ :: IR Eval
until_ = Special untilImpl

untilImpl :: [IR Eval] -> Eval (IR Eval)
untilImpl (cond : body) = loop
  where
    loop = do
        case body of
            [] -> do
                condVal <- eval cond
                case condVal of
                    Bool False -> loop
                    _ -> pure Void
            _ -> do
                mapM_ eval body
                condVal <- eval cond
                case condVal of
                    Bool False -> loop
                    _ -> pure Void
untilImpl _ = throwError $ wrongArgumentType ["condition", "body"]

File: dart/glue/lib/src/lib/bool/until.dart
import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

/// Until loop special form
/// Mirrors Haskell Glue.Lib.Bool.Until.until_ exactly
Eval<Ir> until_(List<Ir> args) {
  return switch (args) {
    [final cond, ...final body] => _loopUntil(cond, body),
    _ => throwError(wrongArgumentType(['condition', 'body'])),
  };
}

/// Internal loop function for until
Eval<Ir> _loopUntil(Ir cond, List<Ir> body) {
  return switch (body) {
    [] => eval(cond).flatMap((condVal) {
      return switch (condVal) {
        IrBool(value: false) => _loopUntil(cond, body),
        _ => Eval.pure(IrVoid()),
      };
    }),
    _ => sequence_(
      body.map(eval).toList(),
      eval(cond).flatMap((condVal) {
        return switch (condVal) {
          IrBool(value: false) => _loopUntil(cond, body),
          _ => Eval.pure(IrVoid()),
        };
      }),
    ),
  };
}
